Add concurrency examples to regression tests

thomasspriggs opened the pull request:
This PR adds concurrency examples to the regression tests in preparation for the planned sequentialization work.
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs commented on regression/cbmc-sequentialization/posix_semaphores/blocking_queue.c:
```diff
@@ -0,0 +1,118 @@
+  enqueue(&input_queue, 8);
+  enqueue(&input_queue, 9);
+  enqueue(&input_queue, 0);
+  join_thread(worker_thread1);
```
I like this example and it should sequentialize well.

martin-cs commented on regression/cbmc-sequentialization/pthread_mutex/no_mutex.c:
```diff
@@ -0,0 +1,52 @@
+  for(int i = 0; i < 100; ++i)
+  {
+    int shared_count_copy = shared_count;
+    // The following call to yield is here in order to cause a thread swap
```
The wording of this sounds a bit certain.  I can believe that 99.99% of the time it does but I don't think it is certain to do so.

martin-cs replied:
I would feel better if the comment was reworded so that it captured this kind of "likely but not certain" but it is not a blocking issue.

hannes-steffenhagen-diffblue replied:
That'd make more sense to me as well, practically this should work but IMHO it’s probably not a good idea to be overly reliant on specific thread scheduling behaviour.

thomasspriggs replied:
I have reworded these comments slightly.

martin-cs commented on regression/cbmc-sequentialization/pthread_mutex/with_mutex.c:
```diff
@@ -0,0 +1,63 @@
+    const int lock_error = pthread_mutex_lock(&mutex);
+    assert(!lock_error);
+    int shared_count_copy = shared_count;
+    // The following call to yield is here in order to cause a thread swap
```
As above.

piotr-grabalski reviewed (approved):
Lots of code but it looks good to me, 

_Personally I am not a big fan of including *.c files though_ ;)

NlightNFotis commented on regression/cbmc-sequentialization/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+  set(gcc_only "-X gcc-only")
```
This is incorrect I believe - it should be: `set(gcc_only -X gcc-only)`. 

The difference is that the code as it stands now passes this as a single string, which passed to `test.pl` creates a problem.

For a more detailed explanation, take a look at the commit message here: https://github.com/diffblue/cbmc/pull/5446/commits/741c303064ab837778087c047b8f1df56d65e549

thomasspriggs replied:
Should be fixed now. Sorry I reintroduced this error, it was a result of cut and pasting from before your fix was merged.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+  set(gcc_only "-X gcc-only")
```
No quotes, this is the same bug fotis was struggling on windows with :)

If you add quotes here this'll be passed as a single argument to test.pl, i.e. `test.pl "-X gcc only"` vs `test.pl -X gcc-only`.

thomasspriggs replied:
Should be fixed now.

thomasspriggs replied:
Actually none of this is going to work on Windows as-is, because pthreads isn't natively supported on that platform. I'll either need to disable all these tests, or add a windows test-time dependency for a suitable compatibility layer.

NlightNFotis commented on regression/cbmc-sequentialization/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+
+add_test_pl_profile(
+    "cbmc-paths-lifo"
+    "$<TARGET_FILE:cbmc> --paths lifo ${gcc_only}"
```
I believe `${gcc_only}` should be outside the quotes here, but I'm not sure.

@hannes-steffenhagen-diffblue ?

thomasspriggs replied:
It was actually on completely the wrong line.

NlightNFotis commented on regression/cbmc-sequentialization/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+
+add_test_pl_profile(
+    "cbmc-cprover-smt2"
+    "$<TARGET_FILE:cbmc> --cprover-smt2 ${gcc_only}"
```
Same as above.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+
+add_test_pl_profile(
+    "cbmc-paths-lifo"
+    "$<TARGET_FILE:cbmc> --paths lifo ${gcc_only}"
```
This gcc_only needs to be in the next line, its an argument to test.pl not cbmc.

thomasspriggs replied:
Should be fixed now.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+
+add_test_pl_profile(
+    "cbmc-cprover-smt2"
+    "$<TARGET_FILE:cbmc> --cprover-smt2 ${gcc_only}"
```
same

thomasspriggs replied:
Should be fixed now.

NlightNFotis reviewed (changes requested):
Can we get changes on the `CMakeLists.txt`. I think they are correctness related and pretty important.

martin-cs reviewed (approved):
Seems good.  Thanks for adding the tests first.  Will you have these run with and without the sequentialisation?  Also, I take it whoever is doing the sequentialisation is aware of the considerable amount of literature on "seqeuntialise and then run CBMC"?

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/pthread_cond/test.c:
```diff
@@ -0,0 +1,82 @@
+}
+
+// Wait for signal. Note that pthreads requires us to lock and unlock the mutex
+// around the call to `pthread_cond_wait`.
```
Maybe worth pointing out if you want to clarify behavior unfamiliar with pthreads that in the period between calling `pthread_cond_wait` and it returning the mutex is actually unlocked, just in case anyone is wondering if this doesn't block everything else.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/pthread_cond/test.c:
```diff
@@ -0,0 +1,82 @@
+{
+  const int lock_error = pthread_mutex_lock(&signal_mutex);
+  assert(!lock_error);
+  const int wait_error = pthread_cond_wait(&condition, &signal_mutex);
```
Writing these comments as I go – note that POSIX requires an associated boolean condition with the condition variable, and correct wake-up behaviour should always be a loop that's waiting and checking the condition after the wait call. Even if the condition you want to wait on is just "someone called signal() at some point" this does, at minimum, still require a flag to operate correctly.

> When using condition variables there is always a Boolean predicate involving shared variables associated with each condition wait that is true if the thread should proceed. Spurious wakeups from the pthread_cond_timedwait() or pthread_cond_wait() functions may occur. Since the return from pthread_cond_timedwait() or pthread_cond_wait() does not imply anything about the value of this predicate, the predicate should be re-evaluated upon such return. 

Without this the expected result for this benchmark should actually be a failure, as a spurious wakeup is possible (this also means that actually we literally don't need to do anything to model `signal` and `wait` correctly other than unlocking and then locking the mutex...). Which to be clear isn't a bad thing, "there is a bug here and we can find it" is a more interesting result (and therefore test) than "we can't find anything here", but this should be documented as such so a future developer doesn't scratch their head why this test is failing despite everything seemingly working as intended.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/pthread_mutex/no_mutex.c:
```diff
@@ -0,0 +1,52 @@
+{
+  for(int i = 0; i < 100; ++i)
+  {
+    int shared_count_copy = shared_count;
```
might be worth pointing out here that the odd construction with the copy-then-increment is because our intended approach (sequentialisation) will give us sequential consistency at the statement level, so an increment operation is going to be atomic for us.

NlightNFotis reviewed (approved):
Thanks for the changes. LGTM! 🎉

thomasspriggs commented:
> Seems good. Thanks for adding the tests first. Will you have these run with and without the sequentialisation? Also, I take it whoever is doing the sequentialisation is aware of the considerable amount of literature on "seqeuntialise and then run CBMC"?

We have been looking into the existing literature and implementations. My thought is to run these specific tests with sequentialisation only, once it is implemented. This is because these test run up against existing limitations without the sequentialisation step. My thought is that the new workflow will look something like -

1. Run goto-cc on the source.
2. Run goto-sequentialise on the goto-binary.
3. Run CBMC against the sequentialised goto-binary. 

Therefore this directory will need some kind of `chain` script, once `goto-sequentialise` exists. Presuming that we keep the sequentialisation in a separate binary.

martin-cs commented:
@thomasspriggs : Some thoughts

- Last I looked CSeq and Yogar-CBMC were the best sequentialisation tools.
- `goto-instrument` already has some concurrent instrumentation (for weak memory), so you could add it there if you want it to implement it as part of a pipeline.
- Most of the sequentialisations I have seen have been underapproximate and slowly increasing the number of interleavings, thread interactions, etc.  It feels like it should even be possible to use the SAT model to tell you where needs more freedom (as with BMC and finding which loops need more unwinding).  As such, you may want to do multiple sequentialisations and have the parameters of the later ones depend on the output of the earlier ones.  If so, do you want to do that with multiple program calls and a script or have one program?

HTH

hannes-steffenhagen-diffblue reviewed (commented):
Not done reviewing yet just to get these out of the way:

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/posix_semaphores/util.c:
```diff
@@ -0,0 +1,33 @@
+
```
I do have to agree the `util.c` is _slightly_ sketchy, generally I prefer if tests stay self contained, even if that means duplicating some code. If you think it's worth it to give more emphasis to the parts that are actually different between tests we can leave it as is though.

thomasspriggs replied:
I have removed `util.c` and duplicated the code. I instinctively dislike cut and pasting code. But I am willing to accept that different ways of doing things may be better for test code.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/posix_semaphores/create_destroy.c:
```diff
@@ -0,0 +1,20 @@
+
+sem_t global_semaphore;
+
+void *worker(void *arguments)
```
:pick: Unused function?

thomasspriggs replied:
Removed.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/posix_semaphores/util.c:
```diff
@@ -0,0 +1,33 @@
+  const pthread_attr_t *const attributes = NULL;
+  void *const worker_argument = NULL;
+  const int create_status =
+    pthread_create(&worker_thread, attributes, &worker, worker_argument);
```
I really don’t like that this assumes the existence of a function with exactly this name, what's wrong with passing this in as an argument?

thomasspriggs replied:
I have now fixed this for the semaphores examples.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/posix_semaphores/blocking_queue.c:
```diff
@@ -0,0 +1,118 @@
+    "enqueue begin:%lu end:%lu data:%d \n", queue->begin, queue->end, data);
+  const int free_wait_error = sem_wait(&queue->free_space);
+  assert(free_wait_error == false);
+  queue->elements[queue->end] = data;
```
:question: 

Is the intention here that `enqueue` and `dequeue` are only allowed to be called from one thread respectively? Otherwise this would be a data race. If so, can you document this here?

thomasspriggs replied:
Yes. This queue implementation is only safe with a single enqueuing and dequeuing thread. I will add some comments to detail this limitation.

thomasspriggs replied:
Added.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/rw_lock/race_test.c:
```diff
@@ -0,0 +1,2 @@
+#define NO_LOCKS
```
Again, I don't like having tests split up over multiple files like this too much if we don't have to..

thomasspriggs replied:
Ok. I have pushed a commit which duplicates this instead.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/rw_lock/with_lock.c:
```diff
@@ -0,0 +1,138 @@
+  struct sharedt *shared = (struct sharedt *)arguments;
+  for(int i = 0; i < 1000; ++i)
+  {
+#ifndef NO_LOCKS
```
I think these would be easier to follow if you had just copy&pasted the code.

thomasspriggs replied:
Done.

hannes-steffenhagen-diffblue commented on regression/cbmc-sequentialization/posix_semaphores/blocking_wait.desc:
```diff
@@ -1,4 +1,4 @@
-FUTURE
+FUTURE requires_posix_only_headers
```
I'd be careful with naming like this, AFAIK Windows is (or at least has been at some point) POSIX compliant.

thomasspriggs replied:
A standard install of Visual Studio does not make the POSIX header files we need available as part of its include path. Without these the build environment is does not have the POSIX compliance level we need. The windows API calls for threading are available through separate header and will have different naming conventions entirely. If a Windows based build environment were to be setup with the appropriate POSIX header files made available, then it should be possible to make these tests work on Windows.

Technically I believe these these would need a minimum of POSIX.1b for semaphores and POSIX.1c for thread creation and joining. The original POSIX.1 version would not be sufficient. But version POSIX.1-2001 onward would have everything we need.

hannes-steffenhagen-diffblue reviewed (approved):
Couple nitpicks here and there, nothing blocking I think.

hannes-steffenhagen-diffblue reviewed (approved):
Thanks for indulging my nitpicks here

thomasspriggs committed:
Add test directory for future sequentialization work

This directory of regression tests is for examples which we intend to
use for testing the future work on the analysis of concurrent programs
based on a sequentialization approach.

Tests which require POSIX specific headers will be disabled for the
moment because these are not natively available on Windows. A windows
POSIX compatability layer library could potentially be used to enable
them to be run on Windows. However the work to do so has not been done.

thomasspriggs committed:
Add pthreads examples

thomasspriggs committed:
Add test using `pthread_rwlock`

thomasspriggs committed:
Duplicate `rw_lock` code

So that each example can be read separately it its own right.

thomasspriggs committed:
Add posix semaphore examples
